File: kindavm/hid.py
"""Core HID report sending functionality."""

import time
from pathlib import Path

HID_DEVICE = "/dev/hidg0"


def send_report(report: bytes, delay_ms: float = 10) -> None:
    """Send a HID report to the device.

    Args:
        report: The HID report bytes to send
        delay_ms: Delay in milliseconds after sending the report
    """
    with open(HID_DEVICE, "rb+") as hid:
        hid.write(report)

    if delay_ms > 0:
        time.sleep(delay_ms / 1000.0)


def check_device() -> bool:
    """Check if the HID device is available.

    Returns:
        True if device exists and is writable, False otherwise
    """
    device = Path(HID_DEVICE)
    return device.exists() and device.is_char_device()

File: internal/hid/hid.go
// Package hid provides HID device interaction for keyboard and mouse control.
package hid

import (
	"fmt"
	"os"
	"time"
)

// Default configuration values
const (
	DefaultHIDDevice = "/dev/hidg0"
	DefaultDelayMS   = 10
)

// Device represents a HID device interface
type Device struct {
	path string
}

// NewDevice creates a new HID device interface
func NewDevice(path string) *Device {
	if path == "" {
		path = DefaultHIDDevice
	}
	return &Device{path: path}
}

// SendReport sends a HID report to the device
func (d *Device) SendReport(report []byte, delayMS int) error {
	f, err := os.OpenFile(d.path, os.O_RDWR, 0o666)
	if err != nil {
		return fmt.Errorf("failed to open HID device: %w", err)
	}
	defer func() {
		_ = f.Close()
	}()

	_, err = f.Write(report)
	if err != nil {
		return fmt.Errorf("failed to write HID report: %w", err)
	}

	if delayMS > 0 {
		time.Sleep(time.Duration(delayMS) * time.Millisecond)
	}

	return nil
}

// CheckDevice verifies that the HID device is available
func (d *Device) CheckDevice() error {
	info, err := os.Stat(d.path)
	if err != nil {
		return fmt.Errorf("HID device not found: %w", err)
	}

	mode := info.Mode()
	if mode&os.ModeCharDevice == 0 {
		return fmt.Errorf("path is not a character device: %s", d.path)
	}

	return nil
}
